### PR TITLE
#1842 Add Currying Design Pattern

### LINKED ISSUE #1842: Currying
I's a functional programming design pattern. As far as I understand you call a function with less parameters than usual and it can be used to defer processing when it has many steps involved.

UML example

```
@startuml
'https://plantuml.com/class-diagram

interface BiFunction <ParamType1, ParamType2, ResultType>{
}

interface Curried<ParamType1, ParamType2 extends Simple> {
  +ParamType2 apply(final ParamType1 a);
}

interface Simple<ParamType1, ParamType2, ResultType> extends BiFunction {
  +ResultType apply(final ParamType1 a, final ParamType2 b);
}

@enduml
```

image

With the curried version you can do stuff like:

```
   System.out.println(curriedAdd.apply(4).apply(5));

    Function1<Integer, Integer> adder5 = curriedAdd.apply(5);
    //some complicate and time consuming process here
    System.out.println(adder5.apply(4));
    System.out.println(adder5.apply(6));

```

### PR DESCRIPTION
**Pull request by hkattt**
This PR resolves #1842.

I adapted the currying example suggested in the #1842 thread. My programmatic example focuses on using a curried builder function to create Book objects.

### CODE REVIEW

**Comment by hkattt** [comment]
Hey @iluwatar, apologies for that. I plan on continuing work on this issue in about 2 weeks time. Sorry for the delay.

File: currying/README.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
+---
+layout: pattern
+title: Currying
+folder: currying
+permalink: /patterns/currying/
+categories: Functional
+language: en
+tags: 
+ - Decoupling
+---
```
Comment (iluwatar) [inline currying/README.md:10]: Please implement the yaml frontmatter according to https://github.com/iluwatar/java-design-patterns/wiki/01.-How-to-contribute

File: currying/src/main/java/com/iluwatar/currying/Book.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  private final String author;
+  private final String title;
+  private final LocalDate publicationDate;
+
+  Book(Genre genre, String author, String title, LocalDate publicationDate) {
+    this.genre = genre;
+    this.author = author;
+    this.title = title;
+    this.publicationDate = publicationDate;
+  }
```
Comment (iluwatar) [inline currying/src/main/java/com/iluwatar/currying/Book.java:45]: Use Lombok to reduce boilerplate

File: currying/src/main/java/com/iluwatar/currying/Genre.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ */
+public enum Genre {
+  FANTASY,
+  HORROR,
+  SCIFI;
+
+  @Override
+  public String toString() {
+    return super.toString();
+  }
```
Comment (iluwatar) [inline currying/src/main/java/com/iluwatar/currying/Genre.java:38]: Doesn't look useful

**Review by iluwatar — CHANGES_REQUESTED** [review]
We need a couple of minor changes. Otherwise, this looks very good!

**Comment by hkattt** [comment]
Hey @iluwatar, I made the changes that you requested; however, the build fails when executing the Maven goal for the service-layer design pattern. Do you have any idea what might be causing this issue?

**Comment by iluwatar** [comment]
Perhaps try to rebase your branch on top of `master` to get rid of the build errors?

**Comment by hkattt** [comment]
Hey @iluwatar, made the changes that you asked for and the tests are now passing.

**Review by iluwatar — APPROVED** [review]

**Comment by iluwatar** [comment]
Excellent, thank you for contributing the new design pattern! 🎉😄💪

@all-contributors please add @hkattt for code

**Comment by iluwatar** [comment]
@all-contributors please add @hkattt for code

**Pull request merged**